File: services/analytics/growthbook.ts
/**
 * services/analytics/growthbook.ts — Mossen 个人版兼容 wrapper (G6-1 + G6-2 重写).
 *
 * 历史:
 *   - 此文件原为 GrowthBook SDK 客户端 (~1228 行), 含远程 init / refresh /
 *     experimentDataByFeature / Statsig fallback / disk cache 等远程能力.
 *   - G6-1: 缩成本地 facade wrapper, 不再 import @growthbook/growthbook
 *   - G6-2: 删除所有远程 init / refresh / reset / processRemoteEvalPayload 内部逻辑
 *   - G6-3: bun remove @growthbook/growthbook (待 Allen 显式确认)
 *   - G6-4: 文案/注释清理
 *
 * 公共 API 保留 22 个 export, 行为完全由 Mossen facade (services/config) 决定:
 *   - getFeatureValue_CACHED_MAY_BE_STALE     → resolveMossenConfig
 *   - getDynamicConfig_CACHED_MAY_BE_STALE    → resolveMossenConfig
 *   - getFeatureValue_CACHED_WITH_REFRESH     → resolveMossenConfig (refresh 参数忽略)
 *   - checkStatsigFeatureGate_CACHED_MAY_BE_STALE → resolveMossenConfig + Boolean
 *   - checkGate_CACHED_OR_BLOCKING            → resolveMossenConfig + Boolean
 *   - getDynamicConfig_BLOCKS_ON_INIT         → resolveMossenConfig (async, 立即返回)
 *   - getFeatureValue_DEPRECATED              → resolveMossenConfig (async)
 *   - checkSecurityRestrictionGate            → 永远返回 false (个人版无 security gate)
 *
 * 远程相关 API 全部 no-op, 保留导出仅为 backward-compat:
 *   - initializeGrowthBook                    → no-op (resolves immediately)
 *   - resetGrowthBook                         → no-op
 *   - refreshGrowthBookFeatures               → no-op
 *   - refreshGrowthBookAfterAuthChange        → no-op
 *   - setupPeriodicGrowthBookRefresh          → no-op
 *   - stopPeriodicGrowthBookRefresh           → no-op
 *   - onGrowthBookRefresh                     → 转发到 onMossenConfigRefresh
 *   - hasGrowthBookEnvOverride                → 检查 MOSSEN_CONFIG_OVERRIDES (新)
 *                                                + MOSSEN_INTERNAL_FC_OVERRIDES (旧 deprecated)
 *   - getAllGrowthBookFeatures                → getAllMossenConfigValues
 *   - getGrowthBookConfigOverrides            → 空对象 (legacy ~/.mossen.json 字段冷冻)
 *   - setGrowthBookConfigOverride             → 转发到 setMossenConfigOverride
 *   - clearGrowthBookConfigOverrides          → 转发到 clearMossenConfigOverrides
 *   - getApiBaseUrlHost                       → utils/customBackend (本地)
 */

import {
  onMossenConfigRefresh,
  resolveMossenConfig,
  setMossenConfigOverride,
  clearMossenConfigOverrides,
  getAllMossenConfigValues,
  resolveAliasedKey,
} from '../config/index.js'
import { getHostedPlatformUrls } from '../../utils/customBackend.js'
import {
  type GitHubActionsMetadata,
} from '../../utils/user.js'

// ============================================================================
// 公共类型 (保留外部调用方可能 import 的 type)
// ============================================================================

/** GrowthBook user attributes; 仅类型保留, 个人版无远程上报 */
export type GrowthBookUserAttributes = {
  id: string
  sessionId: string
  deviceID: string
  platform: 'win32' | 'darwin' | 'linux'
  apiBaseUrlHost?: string
  organizationUUID?: string
  accountUUID?: string
  userType?: string
  subscriptionType?: string
  rateLimitTier?: string
  firstTokenTime?: number
  email?: string
  appVersion?: string
  github?: GitHubActionsMetadata
}

// ============================================================================
// 内部 facade-first helper
// ============================================================================

/**
 * 经 Mossen facade 解析 GrowthBook key.
 *
 * G6-1 重写: 不再 fallback 到远程 GrowthBook. 所有 key 走 facade chain
 * (override > env > project > user > default). MOSSEN_BUILTIN_DEFAULTS 已注入
 * 已迁移 key 的代码默认值 (G3/G4/G5); 未迁移 key fallback 到 caller defaultValue.
 */
function resolveViaFacade<T>(growthbookKey: string, defaultValue: T): T {
  const aliased = resolveAliasedKey(growthbookKey)
  const r = resolveMossenConfig<T>(aliased, defaultValue)
  return r.value
}

// ============================================================================
// Public API: feature value / dynamic config / gate
// ============================================================================

export function getFeatureValue_CACHED_MAY_BE_STALE<T>(
  feature: string,
  defaultValue: T,
): T {
  return resolveViaFacade(feature, defaultValue)
}

/**
 * @deprecated refresh interval 参数被忽略 (Mossen 本地配置无定时刷新需要).
 * 使用 getFeatureValue_CACHED_MAY_BE_STALE.
 */
export function getFeatureValue_CACHED_WITH_REFRESH<T>(
  feature: string,
  defaultValue: T,
  _refreshIntervalMs: number,
): T {
  return resolveViaFacade(feature, defaultValue)
}

export function getDynamicConfig_CACHED_MAY_BE_STALE<T>(
  config: string,
  defaultValue: T,
): T {
  return resolveViaFacade(config, defaultValue)
}

/**
 * G6-2 后行为: 立即解析 (无 init 阻塞), 因为本地 facade 是同步的.
 * 保留 async 签名给 backward-compat.
 */
export async function getDynamicConfig_BLOCKS_ON_INIT<T>(
  config: string,
  defaultValue: T,
): Promise<T> {
  return resolveViaFacade(config, defaultValue)
}

/**
 * @deprecated 远程 GrowthBook 已删除; 仅做本地 facade 解析. 保留 async 签名.
 */
export async function getFeatureValue_DEPRECATED<T>(
  feature: string,
  defaultValue: T,
): Promise<T> {
  return resolveViaFacade(feature, defaultValue)
}

export function checkStatsigFeatureGate_CACHED_MAY_BE_STALE(
  gate: string,
): boolean {
  return Boolean(resolveViaFacade<unknown>(gate, false))
}

/**
 * G6-2 后行为: 同 checkStatsigFeatureGate_CACHED_MAY_BE_STALE 但保留 async 签名.
 */
export async function checkGate_CACHED_OR_BLOCKING(
  gate: string,
): Promise<boolean> {
  return Boolean(resolveViaFacade<unknown>(gate, false))
}

/**
 * Security restriction gate. 个人版永远返回 false (无 hosted security 上报路径).
 */
export async function checkSecurityRestrictionGate(
  _gate: string,
): Promise<boolean> {
  return false
}

// ============================================================================
// 远程相关 API: 全部 no-op (G6-2 删除所有 GrowthBook init/refresh 逻辑)
// ============================================================================

/** No-op: GrowthBook 远程 client 已删. 立即 resolve. */
export const initializeGrowthBook = async (): Promise<void> => {
  // G6-2: 不再初始化远程 client; Mossen facade 是 process-内同步.
}

/** No-op. */
export function resetGrowthBook(): void {
  // G6-2: 远程 client 已删, 无 state 可 reset.
}

/** No-op. */
export async function refreshGrowthBookFeatures(): Promise<void> {
  // G6-2: 无远程刷新.
}

/** No-op. */
export function refreshGrowthBookAfterAuthChange(): void {
  // G6-2: 无 auth-driven refresh.
}

/** No-op. */
export function setupPeriodicGrowthBookRefresh(): void {
  // G6-2: 无 periodic refresh timer.
}

/** No-op. */
export function stopPeriodicGrowthBookRefresh(): void {
  // G6-2: 无 timer 可停.
}

/** Refresh listener — 转发到 Mossen facade refresh listener. */
export function onGrowthBookRefresh(
  listener: () => void | Promise<void>,
): () => void {
  return onMossenConfigRefresh(listener)
}

/**
 * 检查指定 feature 是否被 env override.
 * 个人版只看 MOSSEN_CONFIG_OVERRIDES (新) + MOSSEN_INTERNAL_FC_OVERRIDES (旧 deprecated).
 */
export function hasGrowthBookEnvOverride(feature: string): boolean {
  const newRaw = process.env.MOSSEN_CONFIG_OVERRIDES
  const oldRaw = process.env.MOSSEN_INTERNAL_FC_OVERRIDES
  for (const raw of [newRaw, oldRaw]) {
    if (!raw) continue
    try {
      const parsed = JSON.parse(raw) as Record<string, unknown>
      if (parsed && Object.prototype.hasOwnProperty.call(parsed, feature)) {
        return true
      }
    } catch {
      // ignore malformed JSON
    }
  }
  return false
}

/**
 * 返回当前所有已注入 facade 的 key/value (~Mossen builtin defaults + 任意 override).
 */
export function getAllGrowthBookFeatures(): Record<string, unknown> {
  return getAllMossenConfigValues()
}

/**
 * Legacy 兼容: 返回 ~/.mossen.json 上的 GrowthBook config override 字段.
 * G6-2 后此字段不再被写入; 始终返回空对象.
 */
export function getGrowthBookConfigOverrides(): Record<string, unknown> {
  return {}
}

/** 转发到 setMossenConfigOverride('override' scope). */
export function setGrowthBookConfigOverride(key: string, value: unknown): void {
  setMossenConfigOverride(key, value, 'override')
}

/** 转发到 clearMossenConfigOverrides('override' scope). */
export function clearGrowthBookConfigOverrides(): void {
  clearMossenConfigOverrides('override')
}

/**
 * 当前 API base URL host (来自 customBackend, 本地解析).
 * 个人版用 custom backend (e.g. dashscope), 不走 hosted GrowthBook 端点.
 */
export function getApiBaseUrlHost(): string | undefined {
  try {
    const { remoteBaseUrl } = getHostedPlatformUrls()
    if (!remoteBaseUrl) return undefined
    return new URL(remoteBaseUrl).host
  } catch {
    return undefined
  }
}

File: crates/mossen-agent/src/services/analytics/growthbook.rs
//! GrowthBook compatibility wrapper — local facade for feature flag resolution.
//!
//! All feature value lookups are resolved through the local Mossen config facade.
//! No remote GrowthBook client exists in the personal edition.

use std::collections::HashMap;
use std::env;
use std::sync::{Arc, Mutex, RwLock};

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// GrowthBook user attributes (type retained for backward compat).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct GrowthBookUserAttributes {
    pub id: String,
    pub session_id: String,
    pub device_id: String,
    pub platform: String,
    pub api_base_url_host: Option<String>,
    pub organization_uuid: Option<String>,
    pub account_uuid: Option<String>,
    pub user_type: Option<String>,
    pub subscription_type: Option<String>,
    pub rate_limit_tier: Option<String>,
    pub first_token_time: Option<u64>,
    pub email: Option<String>,
    pub app_version: Option<String>,
}

/// Config override storage.
static CONFIG_OVERRIDES: RwLock<Option<HashMap<String, Value>>> = RwLock::new(None);

/// Resolve a feature value via local facade.
fn resolve_via_facade<T: for<'de> Deserialize<'de> + Clone>(key: &str, default_value: T) -> T {
    // Check env overrides first
    if let Some(val) = get_env_override(key) {
        if let Ok(parsed) = serde_json::from_value::<T>(val) {
            return parsed;
        }
    }
    // Check programmatic overrides
    if let Ok(overrides) = CONFIG_OVERRIDES.read() {
        if let Some(map) = &*overrides {
            if let Some(val) = map.get(key) {
                if let Ok(parsed) = serde_json::from_value::<T>(val.clone()) {
                    return parsed;
                }
            }
        }
    }
    default_value
}

fn get_env_override(key: &str) -> Option<Value> {
    for env_key in &["MOSSEN_CONFIG_OVERRIDES", "MOSSEN_INTERNAL_FC_OVERRIDES"] {
        if let Ok(raw) = env::var(env_key) {
            if let Ok(parsed) = serde_json::from_str::<HashMap<String, Value>>(&raw) {
                if let Some(val) = parsed.get(key) {
                    return Some(val.clone());
                }
            }
        }
    }
    None
}

/// Get feature value (cached, may be stale). Primary public API.
pub fn get_feature_value_cached_may_be_stale<T: for<'de> Deserialize<'de> + Clone>(
    feature: &str,
    default_value: T,
) -> T {
    resolve_via_facade(feature, default_value)
}

/// Get feature value with refresh (refresh interval ignored in personal edition).
pub fn get_feature_value_cached_with_refresh<T: for<'de> Deserialize<'de> + Clone>(
    feature: &str,
    default_value: T,
    _refresh_interval_ms: u64,
) -> T {
    resolve_via_facade(feature, default_value)
}

/// Get dynamic config (cached, may be stale).
pub fn get_dynamic_config_cached_may_be_stale<T: for<'de> Deserialize<'de> + Clone>(
    config: &str,
    default_value: T,
) -> T {
    resolve_via_facade(config, default_value)
}

/// Get dynamic config (blocks on init — in personal edition, resolves immediately).
pub async fn get_dynamic_config_blocks_on_init<T: for<'de> Deserialize<'de> + Clone>(
    config: &str,
    default_value: T,
) -> T {
    resolve_via_facade(config, default_value)
}

/// Deprecated: async feature value getter.
pub async fn get_feature_value_deprecated<T: for<'de> Deserialize<'de> + Clone>(
    feature: &str,
    default_value: T,
) -> T {
    resolve_via_facade(feature, default_value)
}

/// Check a Statsig feature gate (cached, may be stale).
pub fn check_statsig_feature_gate_cached_may_be_stale(gate: &str) -> bool {
    resolve_via_facade::<bool>(gate, false)
}

/// Check gate (cached or blocking — in personal edition, always synchronous).
pub async fn check_gate_cached_or_blocking(gate: &str) -> bool {
    resolve_via_facade::<bool>(gate, false)
}

/// Security restriction gate. Personal edition always returns false.
pub async fn check_security_restriction_gate(_gate: &str) -> bool {
    false
}

/// No-op: GrowthBook remote client removed.
pub async fn initialize_growthbook() {}

/// No-op.
pub fn reset_growthbook() {}

/// No-op.
pub async fn refresh_growthbook_features() {}

/// No-op.
pub fn refresh_growthbook_after_auth_change() {}

/// No-op.
pub fn setup_periodic_growthbook_refresh() {}

/// No-op.
pub fn stop_periodic_growthbook_refresh() {}

/// Check if a feature has an env override.
pub fn has_growthbook_env_override(feature: &str) -> bool {
    get_env_override(feature).is_some()
}

/// Get all feature values (from overrides + defaults).
pub fn get_all_growthbook_features() -> HashMap<String, Value> {
    let overrides = CONFIG_OVERRIDES.read().unwrap();
    overrides.clone().unwrap_or_default()
}

/// Legacy: always returns empty (field frozen).
pub fn get_growthbook_config_overrides() -> HashMap<String, Value> {
    HashMap::new()
}

/// Set a config override.
pub fn set_growthbook_config_override(key: &str, value: Value) {
    let mut overrides = CONFIG_OVERRIDES.write().unwrap();
    let map = overrides.get_or_insert_with(HashMap::new);
    map.insert(key.to_string(), value);
}

/// Clear all config overrides.
pub fn clear_growthbook_config_overrides() {
    let mut overrides = CONFIG_OVERRIDES.write().unwrap();
    *overrides = None;
}

/// Get current API base URL host (from custom backend config).
pub fn get_api_base_url_host() -> Option<String> {
    env::var("CUSTOM_API_BASE_URL")
        .ok()
        .and_then(|url| url::Url::parse(&url).ok())
        .and_then(|u| u.host_str().map(|h| h.to_string()))
}

/// Register a callback invoked when GrowthBook refreshes its values. Mirrors
/// TS `onGrowthBookRefresh(listener)`. Returns an unsubscribe closure.
pub fn on_growth_book_refresh(listener: Box<dyn Fn() + Send + Sync>) -> Box<dyn FnOnce() + Send> {
    // GrowthBook refresh listeners are stored module-level; we keep them
    // alive via a Vec of boxed callbacks.
    let mut guard = GROWTHBOOK_REFRESH_LISTENERS.lock().unwrap();
    guard.push(listener);
    let id = guard.len();
    drop(guard);
    Box::new(move || {
        let mut guard = GROWTHBOOK_REFRESH_LISTENERS.lock().unwrap();
        if id <= guard.len() {
            guard.remove(id - 1);
        }
    })
}

static GROWTHBOOK_REFRESH_LISTENERS: once_cell::sync::Lazy<
    std::sync::Mutex<Vec<Box<dyn Fn() + Send + Sync>>>,
> = once_cell::sync::Lazy::new(|| std::sync::Mutex::new(Vec::new()));
